[eventhubs-checkpointstore-blob] Add README and changelog (#5077)
Azure/azure-sdk-for-js: sdk/eventhub/eventhubs-checkpointstore-blob/README.md

# Azure Event Hubs Checkpoint Store library for Javascript using Storage Blobs

An Azure Blob storage based solution to store checkpoints and to aid in load balancing when using `EventProcessor` from the [@azure/event-hubs](https://www.npmjs.com/package/@azure/event-hubs) library

[Source code](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/eventhub/eventhubs-checkpointstore-blob) | [Package (npm)](https://www.npmjs.com/package/@azure/eventhubs-checkpointstore-blob) | [API Reference Documentation](https://azure.github.io/azure-sdk-for-js/eventhubs-checkpointstore-blob/index.html) | [Samples](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/eventhub/eventhubs-checkpointstore-blob/samples)

## Getting started

### Install the package

Install the Azure Event Hubs Checkpoint Store Blob library using npm

`npm install @azure/eventhubs-checkpointstore-blob@latest`

> **Prerequisites**: You must have an [Azure subscription](https://azure.microsoft.com/free/) and a

Reviewer: ```suggestion
**Prerequisites**: You must have an [Azure subscription](https://azure.microsoft.com/free/), an
```

[Event Hubs Namespace](https://docs.microsoft.com/en-us/azure/event-hubs/) to use this package, and a [Storage account](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction)
If you are using this package in a Node.js application, then use Node.js 8.x or higher.

### Configure Typescript

TypeScript users need to have Node type definitions installed:

```bash
npm install @types/node
```

You also need to enable `compilerOptions.allowSyntheticDefaultImports` in your tsconfig.json. Note that if you have enabled `compilerOptions.esModuleInterop`, `allowSyntheticDefaultImports` is enabled by default. See [TypeScript's compiler options handbook](https://www.typescriptlang.org/docs/handbook/compiler-options.html) for more information.

> ## Key concepts

Reviewer: ```suggestion

Event Processor based application consists of one or more instances of `EventProcessor` which have been
configured to consume events from the same Event Hub and consumer group. They balance the
workload across different instances by distributing the partitions to be processed among themselves.
They also allow the user to track progress when events are processed using checkpoints.
 
 A checkpoint is meant to represent the last successfully processed event by the user from a particular
 partition of a consumer group in an Event Hub instance.
 
A Partition Manager is a class that implements key methods required by the Event Processor to balance load and update checkpoints.
```

> ## Examples

Reviewer: ```suggestion
 ## Examples

### Create a Partition Manager using Azure Blob Storage

To create a Partition Manager that uses Azure Blob Storage, you need to first have an [Azure subscription](https://azure.microsoft.com/free/) and an [Azure Storage account](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction)

Then, you can use the below code snippet to create a Partition Manager

```javascript
import { ContainerClient } from "@azure/storage-blob",
import { BlobPartitionManager } from "@azure/eventhubs-checkpointstore-blob"

const containerClient = new ContainerClient("storage-connection-string", "container-name");
await containerClient.create(); // This can be skipped if the container already exists

const partitionManager =  new BlobPartitionManager(containerClient);
```

> ### Persistent event checkpointing with Azure Blob storage

Reviewer: ```suggestion
### Create a Partition Manager using Azure Blob Storage

To create a Partition Manager that uses Azure Blob Storage, you need to first have an [Azure subscription](https://azure.microsoft.com/free/) and an [Azure Storage account](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction)

Then, you can use the below code snippet to create a Partition Manager

```javascript

### Use Partition Manager with Event Processor
```
import { ContainerClient } from "@azure/storage-blob",
import { BlobPartitionManager } from "@azure/eventhubs-checkpointstore-blob"

const containerClient = new ContainerClient("storage-connection-string", "container-name");
await containerClient.create(); // This can be skipped if the container already exists

const partitionManager =  new BlobPartitionManager(containerClient);
```

> ### Persistent event checkpointing with Azure Blob storage

Reviewer: ```suggestion
### Checkpoint events using Azure Blob storage
```

that you provide, allowing you to focus on business logic while the processor holds responsibility for managing the underlying consumer
operations including checkpointing and load balancing.

If you want to store checkpoints in [Azure Storage Blobs](https://azure.microsoft.com/en-us/services/storage/blobs/), you can provide an instance of [BlobPartitionManager](https://azure.github.io/azure-sdk-for-js/eventhubs-checkpointstore-blob/classes/blobpartitionmanager.html) to your Event Processor. `BlobPartitionManager` uses Storage Blobs to store checkpoints and balance partition load among all instances of Event Processors.

you can see how to use the `EventProcessor` in the below
example, where we use `BlobPartitionManager` that does 
checkpointing in [Azure Storage Blobs](https://azure.microsoft.com/en-us/services/storage/blobs/).

```javascript
const client = new EventHubClient("my-connection-string", "my-event-hub");
const containerClient = new ContainerClient("storage-connection-string", "container-name");
await containerClient.create();

const processor = new EventProcessor(
  EventHubClient.defaultConsumerGroupName,
  client,

> PartitionProcessor,

Reviewer: @chradek, I know we discussed using the PartitionProcessor that will just log, but this will only showcase instantiating classes and not the power of the partition manager.

I want to suggest having 2 examples:
- Create a Partition Manager using Azure Blob Storage
- Checkpoint events using Azure Blob storage

Reviewer: ### Create a Partition Manager using Azure Blob Storage

To create a Partition Manager that uses Azure Blob Storage, you need to first have an [Azure subscription](https://azure.microsoft.com/free/) and an [Azure Storage account](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction)

Then, you can use the below code snippet to create a Partition Manager

```javascript
import { ContainerClient } from "@azure/storage-blob",
import { BlobPartitionManager } from "@azure/eventhubs-checkpointstore-blob"

const containerClient = new ContainerClient("storage-connection-string", "container-name");
await containerClient.create(); // This can be skipped if the container already exists

const partitionManager =  new BlobPartitionManager(containerClient);
```

### Checkpoint events using Azure Blob storage

To checkpoint events received using an EventProcessor to Azure Blob Storage, you will need to pass an instance of the Partition Manager to the Event Processor along with the code to call the `updateCheckpoint()` method.

In this example, we will use a `SamplePartitionProcessor` that extends the `PartitionProcessor` class in order to checkpoint the last event in the batch.

```javascript
import { ContainerClient } from "@azure/storage-blob",
import { BlobPartitionManager } from "@azure/eventhubs-checkpointstore-blob"
import { EventHubClient, PartitionProcessor } from "@azure/event-hubs"

const eventhubClient = new EventHubClient("event-hub-connectionstring")
const containerClient = new ContainerClient("storage-connection-string", "container-name");
await containerClient.create(); // This can be skipped if the container already exists
const partitionManager =  new BlobPartitionManager(containerClient);

SamplePartitionProcessor extends PartitionProcessor  {
   // add code here to override the processevents method to checkpoint the last event in the batch
}

const eventProcessor = new EventProcessor(....)
eventProcessor.start()
```

Author: In `Prerequisites` section we already talked about - "you need to first have an [Azure subscription](https://azure.microsoft.com/free/) and an [Azure Storage account](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction)"

Do we need this here as well?

Reviewer: No, we can skip that part

## Getting started

### Install the package

Install the Azure Event Hubs Checkpoint Store Blob library using npm

`npm install @azure/eventhubs-checkpointstore-blob@latest`

**Prerequisites**: You must have an [Azure subscription](https://azure.microsoft.com/free/), an
[Event Hubs Namespace](https://docs.microsoft.com/en-us/azure/event-hubs/) to use this package, and a [Storage account](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction)
If you are using this package in a Node.js application, then use Node.js 8.x or higher.

### Configure Typescript

TypeScript users need to have Node type definitions installed:

```bash
npm install @types/node
```

You also need to enable `compilerOptions.allowSyntheticDefaultImports` in your tsconfig.json. Note that if you have enabled `compilerOptions.esModuleInterop`, `allowSyntheticDefaultImports` is enabled by default. See [TypeScript's compiler options handbook](https://www.typescriptlang.org/docs/handbook/compiler-options.html) for more information.

> 

Reviewer: There should be a "Key concepts" header here

If you are using this package in a Node.js application, then use Node.js 8.x or higher.

### Configure Typescript

TypeScript users need to have Node type definitions installed:

```bash
npm install @types/node
```

You also need to enable `compilerOptions.allowSyntheticDefaultImports` in your tsconfig.json. Note that if you have enabled `compilerOptions.esModuleInterop`, `allowSyntheticDefaultImports` is enabled by default. See [TypeScript's compiler options handbook](https://www.typescriptlang.org/docs/handbook/compiler-options.html) for more information.

### Key concepts

- **Checkpointing** is a process by which readers mark or commit their position within a partition event sequence. Checkpointing is the responsibility of the consumer and occurs on a per-partition basis within a consumer group. This responsibility means that for each consumer group, each partition reader must keep track of its current position in the event stream, and can inform the service when it considers the data stream complete.

>   If a reader disconnects from a partition, when it reconnects it begins reading at the checkpoint that was previously submitted by the last reader of that partition in that consumer group. When the reader connects, it passes the offset to the event hub to specify the location at which to start reading. In this way, you can use checkpointing to both mark events as "complete" by downstream applications, and to provide resiliency if a failover between readers running on different machines occurs. It is possible to return to older data by specifying a lower offset from this checkpointing process. Through this mechanism, checkpointing enables both failover resiliency and event stream replay..

Reviewer 2: Nit: Last sentence should end with a single period.

configured to consume events from the same Event Hub and consumer group. They balance the
workload across different instances by distributing the partitions to be processed among themselves.
They also allow the user to track progress when events are processed using checkpoints.
 
  A checkpoint is meant to represent the last successfully processed event by the user from a particular
  partition of a consumer group in an Event Hub instance.
 
  A Partition Manager is a class that implements key methods required by the Event Processor to balance load and update checkpoints.

## Examples

- [Create a Partition Manager using Azure Blob Storage](#create-a-partition-manager-using-azure-blob-storage)
- [Checkpoint events using Azure Blob storage](#checkpoint-events-using-azure-blob-storage)

### Create a Partition Manager using Azure Blob Storage

 you can use the below code snippet to create a Partition Manager

```javascript
import { ContainerClient } from "@azure/storage-blob",
import { BlobPartitionManager } from "@azure/eventhubs-checkpointstore-blob"

const containerClient = new ContainerClient("storage-connection-string", "container-name");

> await containerClient.create(); // This can be skipped if the container already exists

Reviewer 2: Good comment!

const partitionManager =  new BlobPartitionManager(containerClient);
```

### Checkpoint events using Azure Blob storage

To checkpoint events received using an EventProcessor to Azure Blob Storage, you will need to pass an instance of the Partition Manager to the Event Processor along with the code to call the `updateCheckpoint()` method.

> In this example, we will use a `SamplePartitionProcessor` that extends the `PartitionProcessor` class in order to checkpoint the last event in the batch.

Reviewer 2: This example says we'll checkpoint the last event in the batch, but we aren't actually showing the code to do that. Can you include something like 
```typescript
processEvents(events, partitionContext) {
  if (events.length) {
    /* custom logic for processing events, then checkpoint*/
    await partitionContext.updateCheckpoint(events[events.length - 1]);
  }
}
```

## Getting started

### Install the package

Install the Azure Event Hubs Checkpoint Store Blob library using npm

`npm install @azure/eventhubs-checkpointstore-blob`

**Prerequisites**: You must have an [Azure subscription](https://azure.microsoft.com/free/), an
[Event Hubs Namespace](https://docs.microsoft.com/en-us/azure/event-hubs/) to use this package, and a [Storage account](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction)
If you are using this package in a Node.js application, then use Node.js 8.x or higher.

### Configure Typescript

TypeScript users need to have Node type definitions installed:

```bash
npm install @types/node
```

You also need to enable `compilerOptions.allowSyntheticDefaultImports` in your tsconfig.json. Note that if you have enabled `compilerOptions.esModuleInterop`, `allowSyntheticDefaultImports` is enabled by default. See [TypeScript's compiler options handbook](https://www.typescriptlang.org/docs/handbook/compiler-options.html) for more information.

### Key concepts

> - **Checkpointing** is a process by which readers mark or commit their position within a partition event sequence. Checkpointing is the responsibility of the consumer and occurs on a per-partition basis within a consumer group. This responsibility means that for each consumer group, each partition reader must keep track of its current position in the event stream, and can inform the service when it considers the data stream complete.

Reviewer: If we are going to bolden checkpointing, then we should also bolden "load balancing" as these are the 2 features that the Partition Manager is helping out with

  If a reader disconnects from a partition, when it reconnects it begins reading at the checkpoint that was previously submitted by the last reader of that partition in that consumer group. When the reader connects, it passes the offset to the event hub to specify the location at which to start reading. In this way, you can use checkpointing to both mark events as "complete" by downstream applications, and to provide resiliency if a failover between readers running on different machines occurs. It is possible to return to older data by specifying a lower offset from this checkpointing process. Through this mechanism, checkpointing enables both failover resiliency and event stream replay.

- Event Processor based application consists of one or more instances of `EventProcessor` which have been
  configured to consume events from the same Event Hub and consumer group. They balance the
  workload across different instances by distributing the partitions to be processed among themselves.
  They also allow the user to track progress when events are processed using checkpoints.
  A checkpoint is meant to represent the last successfully processed event by the user from a particular

>   partition of a consumer group in an Event Hub instance.

Reviewer: We dont need to talk about what a checkpoint is here again after we added 2 big paragraphs for the same before :)

  A Partition Manager is a class that implements key methods required by the Event Processor to balance load and update checkpoints.

## Examples

- [Create a Partition Manager using Azure Blob Storage](#create-a-partition-manager-using-azure-blob-storage)
- [Checkpoint events using Azure Blob storage](#checkpoint-events-using-azure-blob-storage)

### Create a Partition Manager using Azure Blob Storage

> you can use the below code snippet to create a Partition Manager

Reviewer: ```suggestion
Use the below code snippet to create a Partition Manager. You will need to provide the connection string to your storage account.
```

In this example, we will use a `SamplePartitionProcessor` that extends the [PartitionProcessor](https://azure.github.io/azure-sdk-for-js/event-hubs/classes/partitionprocessor.html) class in order to checkpoint the last event in the batch.

```javascript
import { ContainerClient } from "@azure/storage-blob",
import { BlobPartitionManager } from "@azure/eventhubs-checkpointstore-blob"
import { EventHubClient, PartitionProcessor } from "@azure/event-hubs"

const eventhubClient = new EventHubClient("event-hub-connectionstring")
const containerClient = new ContainerClient("storage-connection-string", "container-name");
await containerClient.create(); // This can be skipped if the container already exists
const partitionManager =  new BlobPartitionManager(containerClient);

SamplePartitionProcessor extends PartitionProcessor  {
  async processEvents(events, partitionContext) {
  if (events.length) {
    /* custom logic for processing events, then checkpoint*/
    await partitionContext.updateCheckpoint(events[events.length - 1]);
  }
}

> }

Reviewer: The indentation is not right here. Can we fix it?

> # Azure Checkpoint Store client library for Javascript using Storage Blobs

Reviewer: ```suggestion
# Azure Event Hubs Checkpoint Store library for Javascript using Storage Blobs
```

> An Azure Blob storage based solution to store checkpoints when using Event Hubs.

Reviewer: ```suggestion
An Azure Blob storage based solution to store checkpoints and to aid in load balancing when using `EventProcessor` from the [@azure/event-hubs](https://www.npmjs.com/package/@azure/event-hubs) library
```

> [Source code](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/eventhub/eventhubs-checkpointstore-blob) | [Package (npm)](https://www.npmjs.com/package/@azure/eventhubs-checkpointstore-blob) | [API Reference Documentation](https://azure.github.io/azure-sdk-for-js/eventhubs-checkpointstore-blob/index.html) | [Product documentation](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-event-processor-host) | [Samples](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/eventhub/eventhubs-checkpointstore-blob/samples)

Reviewer: For now, skip the Product Documentation

## Getting started

### Install the package

> Install the Azure Event Hubs checkpointstore blob library using npm

Reviewer: ```suggestion
Install the Azure Event Hubs Checkpoint Store Blob library using npm
```

`npm install @azure/eventhubs-checkpointstore-blob@1.0.0-preview.1`

**Prerequisites**: You must have an [Azure subscription](https://azure.microsoft.com/free/) and a

> [Event Hubs Namespace](https://docs.microsoft.com/en-us/azure/event-hubs/) to use this package.

Reviewer: ```suggestion
[Event Hubs Namespace](https://docs.microsoft.com/en-us/azure/event-hubs/) to use this package, and a [Storage account](Add link here)
```

If you are using this package in a Node.js application, then use Node.js 8.x or higher.

### Configure Typescript

TypeScript users need to have Node type definitions installed:

```bash
npm install @types/node
```

You also need to enable `compilerOptions.allowSyntheticDefaultImports` in your tsconfig.json. Note that if you have enabled `compilerOptions.esModuleInterop`, `allowSyntheticDefaultImports` is enabled by default. See [TypeScript's compiler options handbook](https://www.typescriptlang.org/docs/handbook/compiler-options.html) for more information.

## Key concepts

## Examples

> - [Consume events from all Event Hub partitions](#consume-events-from-all-event-hub-partitions)

Reviewer 2: It might be more informative to a user if the title was something like
> Persistent event checkpointing with Azure Blob storage
this way they get a clearer idea what the package is for (with the current title I'm thinking I somehow use this package to consume events).

Reviewer: Don't need links to samples if there is only one sample

### Consume events from all Event Hub partitions

> To consume events from all the partitions of an Event Hub, you'll create an [EventProcessor](https://azure.github.io/azure-sdk-for-js/event-hubs/classes/eventprocessor.html)

Reviewer 2: Should this sample live in the samples directory? I'd expect to see a sample of how to construct the PartitionManager in this package, and maybe even using it with `EventProcessor`

Author: We decided that we already have a sample in Event Hubs SDK, so we don't need it here, but I'll add one inside `samples` directory.

Reviewer 2: Do you mean it's been decided to remove this sample altogether from this package?

If so, I think it'd be useful to have a sample that shows how to construct the class from this package, but otherwise agree there's a lot of overlap with the event-hubs samples.

example, where we use `BlobPartitionManager` that does 
checkpointing in [Azure Storage Blobs](https://azure.microsoft.com/en-us/services/storage/blobs/).

```javascript
class SamplePartitionProcessor extends PartitionProcessor {
  // Gets called once before the processing of events from current partition starts.
  async initialize(partitionContext) {
    /* your code here */
  }

  // Gets called for each batch of events that are received.
  // You may choose to use the checkpoint manager to update checkpoints.
  async processEvents(events, partitionContext) {
    /* your code here */
  }

  // Gets called for any error when receiving events.
  async processError(error, partitionContext) {
    /* your code here */
  }

  // Gets called when Event Processor stops processing events for current partition.
  async close(reason, partitionContext) {
    /* your code here */
  }
}

const client = new EventHubClient("my-connection-string", "my-event-hub");
const containerClient = new ContainerClient("storage-connection-string", "container-name");
await containerClient.create();

const processor = new EventProcessor(
  EventHubClient.defaultConsumerGroupName,
  client,

>   SamplePartitionProcessor,

Reviewer: ```suggestion
PartitionProcessor,
```

This way, you can remove the code for SamplePartitionProcessor

  new BlobPartitionManager(containerClient)
);
await processor.start();
// At this point, the processor is consuming events from each partition of the Event Hub and
// delegating them to the SamplePartitionProcessor instance created for that partition.  This
// processing takes place in the background and will not block.
//
// In this example, we'll stop processing after five seconds.
await delay(5000);
await processor.stop();
```

## Troubleshooting

> ### AMQP Dependencies

Author: Not sure, but If user wants to enable the protocol level logs then I think we should mention about AMQP.

> ### AMQP Dependencies

Reviewer 2: Wouldn't they get that info from the event-hubs package?

> ### AMQP Dependencies

Author: Yes but I added the same info here as well.

The Event Hubs library depends on the [rhea-promise](https://github.com/amqp/rhea-promise) library for managing connections, sending and receiving events over the [AMQP](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-complete-v1.0-os.pdf) protocol.

### Enable logs

You can set the following environment variable to get the debug logs when using this library.

- Getting debug logs from the Eventhubs Checkpointstore Blob

```bash

> export DEBUG=azure*

Author: Yes, they can set `azure:eventhubs-checkpointstore-blob:info` but I followed the same structure as Event Hubs.

> export DEBUG=azure*

Reviewer 2: `azure:eventhubs-checkpointstore-blob*` should cover all our bases for this package right now.

I wouldn't use Event Hubs as the gold standard; it has a ton of namespaces which makes it difficult to get all logs for the package unless you get all logs for all Azure SDKs. We don't have that issue with this package.

const partitionManager =  new BlobPartitionManager(containerClient);

SamplePartitionProcessor extends PartitionProcessor  {

>    processEvents(events, partitionContext) {

Reviewer 2: Sorry, minor, but can you make this `async`? That way it matches the expected signature
